File: modules/cubex-core/src/main/java/org/cubexmc/core/CubexText.java
package org.cubexmc.core;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import net.md_5.bungee.api.ChatColor;

public final class CubexText {

    private static final Pattern HEX_PATTERN = Pattern.compile("&#([A-Fa-f0-9]{6})");

    public String color(String input) {
        String colored = colorOrNull(input);
        return colored == null ? "" : colored;
    }

    public String colorOrNull(String input) {
        if (input == null) {
            return null;
        }
        if (input.isEmpty()) {
            return input;
        }

        Matcher matcher = HEX_PATTERN.matcher(input);
        StringBuffer buffer = new StringBuffer();
        while (matcher.find()) {
            matcher.appendReplacement(buffer, ChatColor.of("#" + matcher.group(1)).toString());
        }
        matcher.appendTail(buffer);
        return org.bukkit.ChatColor.translateAlternateColorCodes('&', buffer.toString());
    }

    public String stripControl(String input) {
        if (input == null) {
            return "";
        }
        return input.replaceAll("[\\p{Cntrl}&&[^\r\n\t]]", "").trim();
    }

    public String nullToEmpty(String input) {
        return input == null ? "" : input;
    }
}

File: modules/cubex-core/src/main/kotlin/org/cubexmc/core/CubexText.kt
package org.cubexmc.core

import java.util.regex.Pattern
import net.md_5.bungee.api.ChatColor

class CubexText {
    fun color(input: String?): String = colorOrNull(input) ?: ""

    fun colorOrNull(input: String?): String? {
        if (input == null || input.isEmpty()) return input

        val matcher = HEX_PATTERN.matcher(input)
        val buffer = StringBuffer()
        while (matcher.find()) {
            val replacement = try {
                ChatColor.of("#${matcher.group(1)}").toString()
            } catch (_: NoSuchMethodError) {
                ""
            } catch (_: NoClassDefFoundError) {
                ""
            }
            matcher.appendReplacement(buffer, replacement)
        }
        matcher.appendTail(buffer)
        return org.bukkit.ChatColor.translateAlternateColorCodes('&', buffer.toString())
    }

    fun stripControl(input: String?): String =
        input?.replace(CONTROL_PATTERN, "")?.trim() ?: ""

    fun nullToEmpty(input: String?): String = input ?: ""

    companion object {
        private val SHARED = CubexText()
        private val HEX_PATTERN: Pattern = Pattern.compile("&#([A-Fa-f0-9]{6})")
        private val CONTROL_PATTERN = Regex("[\\p{Cntrl}&&[^\\r\\n\\t]]")

        @JvmStatic
        fun translateColorCodes(input: String?): String? = SHARED.colorOrNull(input)
    }
}
